mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java

```diff
@@ -33,8 +44,111 @@ public DomainBundleProjectValidator(Path projectBaseDir, Project dependencyProje
     this.muleMavenPluginClient = aetherMavenClient;
   }
 
+  /**
+   * Validates that the mule domain dependencies resolution conform to the mule domain bundle package definition.
+   *
+   * @throws ValidationException if the dependency resolution does not conform to the mule domain bundle package definition.
+   */
   @Override
   protected void additionalValidation() throws ValidationException {
-    // TODO MMP-224
+    Set<ArtifactCoordinates> domains =
+        dependencyProject.getDependencies().stream().filter(d -> MULE_DOMAIN.equals(d.getClassifier()))
+            .collect(Collectors.toSet());
+
+    validateDomain(domains);
+
+    List<ArtifactCoordinates> applications =
+        dependencyProject.getDependencies().stream().filter(d -> !MULE_DOMAIN.equals(d.getClassifier()))
+            .collect(Collectors.toList());
+
+    validateApplications(domains.iterator().next(), applications);
+  }
+
+  /**
+   * Validates that all applications refers to a unique domain.
+   *
+   * @param domain maven coordinates of the unique domain
+   * @param applications list of domain bundle applications
+   */
+  protected void validateApplications(ArtifactCoordinates domain, List<ArtifactCoordinates> applications)
+      throws ValidationException {
+    if (applications == null || applications.size() == 0) {
+      throw new ValidationException("A domain bundle should contain at least one application");
+    }
+    for (ArtifactCoordinates applicationCoordinates : applications) {
+      validateApplication(domain, applicationCoordinates);
+    }
+  }
+
+  /**
+   * Validates that an application refers to a unique domain.
+   *
+   * @param domain maven coordinates of the unique domain
+   * @param applicationCoordinates maven coordinates of the application that is being validated
+   */
+  protected void validateApplication(ArtifactCoordinates domain, ArtifactCoordinates applicationCoordinates)
+      throws ValidationException {
+    Set<ArtifactCoordinates> applicationDomains = getApplicationDomains(applicationCoordinates);
+
+    if (applicationDomains.size() != DOMAIN_BUNDLE_VALID_NUMBER_OF_DOMAINS || !applicationDomains.contains(domain)) {
+      throw new ValidationException("Every application in the domain bundle must refer to the specified domain: " + domain
+          + ". However, the application refers to the following domain(s): "
+          + applicationDomains.stream().collect(Collectors.toList()));
```

Comment:
Can we custom the message to reflect more specifically that the application has reference to no domain?

Reply:
Done